### PR TITLE
AUTH-421 Fargate FastAPI container

### PR DESCRIPTION
**Pull request by y3rsh**
## Switch architecture yet again

- We now have 176 seconds to respond to requests.
- Auth is a part of the API.
- Docs <https://staging.opentrons.ai/api/doc>  or <https://staging.opentrons.ai/api/redoc>

This is all deployed on staging and is responding to requests from #15193

### CODE REVIEW

File: opentrons-ai-server/api/handler/fast.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    Access_Control_Expose_Headers: str
+    Access_Control_Max_Age: str
+
+
+# Endpoint functions
+@app.post(
+    CHAT_COMPLETION_ROUTE,
+    response_model=Union[ChatResponse, ErrorResponse],
+    summary="Create Chat Completion",
+    description="Generate a chat response based on the provided message.",
```
Comment (koji) [inline opentrons-ai-server/api/handler/fast.py]: nit

```suggestion
    description="Generate a chat response based on the provided prompt.",
```

File: opentrons-ai-server/tests/test_live.py [hunk 2]
```diff
@@ -29,27 +28,28 @@ def test_get_chat_completion_bad_auth(client: Client) -> None:
...
-    assert response.status_code == 404, "nonexistent endpoint with good auth should return HTTP 404"
+    assert response.status_code == 405, "nonexistent endpoint with good auth should return HTTP 405"
 
 
 @pytest.mark.live
 def test_get_bad_endpoint_with_bad_auth(client: Client) -> None:
     """Test a nonexistent endpoint with bad authentication."""
     response = client.get_bad_endpoint(bad_auth=True)
-    assert response.status_code == 401, "nonexistent endpoint with bad auth should return HTTP 401"
+    assert response.status_code == 405, "nonexistent endpoint with bad auth should return HTTP 405"
```
Comment (koji) [inline opentrons-ai-server/tests/test_live.py]: why 405?

https://developer.mozilla.org/en-US/docs/web/http/status/405

File: opentrons-ai-server/tests/test_live.py [hunk 3]
```diff
@@ -29,27 +28,28 @@ def test_get_chat_completion_bad_auth(client: Client) -> None:
 def test_get_bad_endpoint_with_good_auth(client: Client) -> None:
     """Test a nonexistent endpoint with good authentication."""
     response = client.get_bad_endpoint()
-    assert response.status_code == 404, "nonexistent endpoint with good auth should return HTTP 404"
+    assert response.status_code == 405, "nonexistent endpoint with good auth should return HTTP 405"
```
Comment (koji) [inline opentrons-ai-server/tests/test_live.py]: I think 404 would be better.

**Review by koji — APPROVED** [review]
left a few comments but they aren't a blocker.
the changes make sense to me.
Thank you for doing this!

File: opentrons-ai-server/api/models/chat_request.py [hunk 4]
```diff
@@ -1,6 +1,17 @@
...
 
 
 class ChatRequest(BaseModel):
-    message: str
-    fake: bool
+    message: str = Field(..., description="The latest message to be processed.")
+    history: Optional[List[ChatCompletionMessageParam]] = Field(
+        None, description="Chat history in the form of a list of messages.  Type is from OpenAI's ChatCompletionMessageParam"
+    )
+    fake: bool = Field(True, description="When set to true, the response will be a fake. OpenAI API is not used.")
```
Comment (DerekMaggio) [inline opentrons-ai-server/api/models/chat_request.py:17]: Why would we be making fake requests? Is this for testing all our code that uses OpenAI, without actually calling to it?

**Comment by Elyorcv** [comment]
I want to understand more on 176s. Sometimes, to generate protocol we may need more than 3 mins, do you think 176s is related to this?

File: opentrons-ai-server/README.md [hunk 5]
```diff
@@ -32,30 +34,25 @@ The Opentrons AI application's server.
 
 `python -m pipenv install openai==1.25.1`
```
Comment (koji) [inline opentrons-ai-server/README.md]:
```suggestion
`python -m pipenv install openai==1.30.4`
```

Comment (y3rsh) [reply opentrons-ai-server/README.md]: Fixed

File: opentrons-ai-server/api/handler/fast.py (hunk 1 above)
Comment (y3rsh) [reply opentrons-ai-server/api/handler/fast.py]: Fixed

File: opentrons-ai-server/api/models/chat_request.py (hunk 4 above)
Comment (y3rsh) [reply opentrons-ai-server/api/models/chat_request.py:17]: Correct. A flag for testing to validate the API is up and you may make an authenticated API request without running all the code and pushing a request to OpenAI.

File: opentrons-ai-server/tests/test_live.py (hunk 3 above)
Comment (y3rsh) [reply opentrons-ai-server/tests/test_live.py]: changed.

File: opentrons-ai-server/tests/test_live.py (hunk 2 above)
Comment (y3rsh) [reply opentrons-ai-server/tests/test_live.py]: fixed

**Comment by y3rsh** [comment]
> I want to understand more on 176s. Sometimes, to generate protocol we may need more than 3 mins, do you think 176s is related to this?

178 seconds is the maximum possible for this architecture and is already not optimal for waiting on a POST request to return a response.  There are many other ways to do this we will need to iterate on performance and other methods getting the response.

File: opentrons-ai-server/README.md [hunk 6]
```diff
@@ -20,42 +20,41 @@ The Opentrons AI application's server.
...
+1. docker
 
 ## Install a dev dependency
 
 `python -m pipenv install pytest==8.2.0 --dev`
 
 ## Install a production dependency
 
-`python -m pipenv install openai==1.25.1`
+`python -m pipenv install openai==1.30.4`
```
Comment (Elyorcv) [inline opentrons-ai-server/README.md:35]: We should not keep changing the version please. Reproducibility is already difficult due to GPT model itself.

File: opentrons-ai-server/Pipfile [hunk 7]
```diff
@@ -4,26 +4,29 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-openai = "==1.25.1"
+openai = "==1.30.4"
```
Comment (Elyorcv) [inline opentrons-ai-server/Pipfile:7]: better to stick to 1.25.1, AI projects are updated fast. it may damage things unexpectedly.

**Pull request merged**